core/src/main/scala/fs2/async/mutable/Signal.scala

```diff
@@ -70,18 +70,23 @@ object Signal {
 
       def discrete: Stream[F, A] = {
         def go(lastA:A, last:Long):Stream[F,A] = {
-          def getNext:F[(A,Long)] = {
+          def getNext:F[(F[(A,Long)],F[Unit])] = {
             F.bind(F.ref[(A,Long)]) { ref =>
-              F.bind(F.modify(state) { case s@(a, l, listen) =>
+              F.map(F.modify(state) { case s@(a, l, listen) =>
                 if (l != last) s
                 else (a, l, listen :+ ref)
               }) { c =>
-                if (c.modified) F.get(ref)
-                else F.pure(c.now._1 -> c.now._2)
+                if (c.modified) {
+                  val cleanup = F.map(F.modify(state) {
+                    case s@(a, l, listen) => if (l != last) s else (a, l, listen.filterNot(_ == ref))
+                  })(_ => ())
```

Comment:
Definitely just leave as is.

On Fri, Jun 24, 2016 at 2:50 PM Michael Pilquist <notifications@github.com>
wrote:

> In core/src/main/scala/fs2/async/mutable/Signal.scala
> <https://github.com/functional-streams-for-scala/fs2/pull/662#discussion_r68444881>
> :
>
> >                  if (l != last) s
> >                  else (a, l, listen :+ ref)
> >                }) { c =>
> > -                if (c.modified) F.get(ref)
> > -                else F.pure(c.now._1 -> c.now._2)
> > +                if (c.modified) {
> > +                  val cleanup = F.map(F.modify(state) {
> > +                    case s@(a, l, listen) => if (l != last) s else (a, l, listen.filterNot(_ == ref))
> > +                  })(_ => ())
>
> Ha... so what do you think? Use eq and change Async to define type Ref[A]
> <: AnyRef? Or just leave it as is, using ==?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/functional-streams-for-scala/fs2/pull/662/files/7d1323b1354a85489f5a0c2ee97ae46fd167644b#r68444881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAArQt3uKGs6ZJkXm6as7Z19-OIU1B6Iks5qPCcRgaJpZM4I95P0>
> .
>